engine/src/main/java/com/datatorrent/stram/plan/logical/LogicalPlan.java

```diff
@@ -778,32 +799,162 @@ public void resetStreamPersistanceOnSinkRemoval(InputPortMeta sinkBeingRemoved)
+
+      @Override
+      public void addOutputPort(OutputPort<?> portObject, Field field, OutputPortFieldAnnotation portAnnotation, AppData.ResultPort adrAnnotation)
+      {
+        if (!DagNodeMeta.this.outputStreams.isEmpty()) {
+          for (Map.Entry<LogicalPlan.OutputPortMeta, LogicalPlan.StreamMeta> e : DagNodeMeta.this.outputStreams.entrySet()) {
+            LogicalPlan.OutputPortMeta pm = e.getKey();
+            if (pm.operatorMeta == DagNodeMeta.this && pm.fieldName.equals(field.getName())) {
+              //LOG.debug("Found existing port meta for: " + field);
+              outPortMap.put(portObject, pm);
+              markOutputPortIfHidden(pm.getPortName(), pm, field.getDeclaringClass());
+              return;
+            }
+          }
+        }
+        OutputPortMeta metaPort = new OutputPortMeta();
+        metaPort.operatorMeta = DagNodeMeta.this;
+        metaPort.fieldName = field.getName();
+        metaPort.portAnnotation = portAnnotation;
+        metaPort.adrAnnotation = adrAnnotation;
+        outPortMap.put(portObject, metaPort);
+        markOutputPortIfHidden(metaPort.getPortName(), metaPort, field.getDeclaringClass());
+      }
+
+      private void markOutputPortIfHidden(String portName, OutputPortMeta portMeta, Class<?> declaringClass)
+      {
+        if (!portNameMap.containsKey(portName)) {
+          portNameMap.put(portName, portMeta);
+        } else {
+          // make the port optional
+          portMeta.classDeclaringHiddenPort = declaringClass;
+        }
+
+      }
+
+      private void markInputPortIfHidden(String portName, InputPortMeta portMeta, Class<?> declaringClass)
+      {
+        if (!portNameMap.containsKey(portName)) {
+          portNameMap.put(portName, portMeta);
+        } else {
+          // make the port optional
+          portMeta.classDeclaringHiddenPort = declaringClass;
+        }
+      }
+    }
+
+    /**
+     * Ports objects are transient, we keep a lazy initialized mapping
+     */
+    private transient PortMapping portMapping = null;
+
+    protected PortMapping getPortMapping()
+    {
+      if (this.portMapping == null) {
+        this.portMapping = new PortMapping();
+        Operators.describe(this.getNode(), portMapping);
+      }
+      return portMapping;
+    }
+
+    public OutputPortMeta getMeta(Operator.OutputPort<?> port)
+    {
+      return getPortMapping().outPortMap.get(port);
+    }
+
+    public InputPortMeta getMeta(Operator.InputPort<?> port)
+    {
+      return getPortMapping().inPortMap.get(port);
+    }
+
+    public Map<InputPortMeta, StreamMeta> getInputStreams()
+    {
+      return this.inputStreams;
+    }
+
+    public Map<OutputPortMeta, StreamMeta> getOutputStreams()
+    {
+      return this.outputStreams;
+    }
+
+  }
+
   /**
    * Operator meta object.
    */
-  public final class OperatorMeta implements DAG.OperatorMeta, Serializable
+  public class OperatorMeta extends DagNodeMeta implements DAG.OperatorMeta, Serializable
```

Comment:
Not clear why we need this distinction and why some of the fields are present here and not in "DagNodeMeta".